OpenRA.Mods.Common/Traits/World/TerrainGeometryOverlay.cs

```diff
@@ -51,12 +51,35 @@ void IChatCommand.InvokeCommand(string name, string arg)
 
 		IEnumerable<IRenderable> IRenderAnnotations.RenderAnnotations(Actor self, WorldRenderer wr)
 		{
-			if (!Enabled)
-				yield break;
-
 			var map = wr.World.Map;
 			var colors = wr.World.Map.Rules.TerrainInfo.HeightDebugColors;
 			var mouseCell = wr.Viewport.ViewToWorld(Viewport.LastMousePos).ToMPos(wr.World.Map);
+			var world = wr.Viewport.ViewToWorldPx(Viewport.LastMousePos);
+
+			foreach (var uv in wr.Viewport.CandidateMouseoverCells(world))
+			{
+				var height = (int)map.Height[uv];
+				var r = map.Grid.Ramps[map.Ramp[uv]];
+				var pos = map.CenterOfCell(uv.ToCPos(map)) - new WVec(0, 0, r.CenterHeightOffset);
+				var width = uv == mouseCell ? 3 : 1;
+
+				// Colors change between points, so render separately
+				foreach (var p in r.Polygons)
+				{
+					for (var i = 0; i < p.Length; i++)
+					{
+						var j = (i + 1) % p.Length;
+						var start = pos + p[i];
+						var end = pos + p[j];
+						var startColor = colors[height + p[i].Z / 512];
+						var endColor = colors[height + p[j].Z / 512];
+						yield return new LineAnnotationRenderable(start, end, width, startColor, endColor);
+					}
+				}
+			}
+
```

Comment:
Wouldn't this cause a large performanc degredation if above code is always executed even if the overlay is disabled?

Reply:
this is the testcase commit, it's not meant to be merged